### PR TITLE
add encoding in client functions

### PR DESCRIPTION
**Pull request by JianGuoPinterest**
in order to support remcached operations, I extracted the key helper function, it won't affect the current logic, this _key_helper function will be override in the remcached client implementation.

### CODE REVIEW

File: pymemcache/client/base.py [hunk 1]
```diff
@@ -739,6 +739,10 @@ def _raise_errors(self, line, name):
             error = line[line.find(b' ') + 1:]
             raise MemcacheServerError(error)
 
+    # extract the key helper to support extensions
```
Comment (jparise) [inline pymemcache/client/base.py]: Because we're doing this to support possible extension via subclasses, could you add a little bit more documentation here so that the contract is clearer to end users? I think using a docblock would probably be better than a comment for that purpose, too.

File: pymemcache/client/base.py [hunk 2]
```diff
@@ -304,6 +305,9 @@ def set(self, key, value, expire=0, noreply=None):
                   from the cache, or zero for no expiry (the default).
           noreply: optional bool, True to not wait for the reply (defaults to
                    self.default_noreply).
+          flag: remcached operation flag
```
Comment (LINKIWI) [inline pymemcache/client/base.py]: The `flag` is not specific to remcached, but rather [one component](https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L149) of the memcached protocol's underlying storage command API. Let's update the documentation to reflect this.

File: pymemcache/client/base.py [hunk 3]
```diff
@@ -304,6 +305,9 @@ def set(self, key, value, expire=0, noreply=None):
...
+          encoding: pymemcached use ascii, remcached use utf-8
```
Comment (LINKIWI) [inline pymemcache/client/base.py]: Again there is no need to reference remcached here as this is a generic, public-facing API. The encoding value here controls how the data is encoded before shipping it on the wire. The fact that remcached chooses `utf-8` is an implementation detail.

File: pymemcache/client/base.py [hunk 4]
```diff
@@ -739,6 +744,22 @@ def _raise_errors(self, line, name):
             error = line[line.find(b' ') + 1:]
             raise MemcacheServerError(error)
 
+    def _key_helper(self, key, remapped_keys, prefixed_keys):
+        """
+        This function is extracted from _fetch_cmd to support
+            future extension for remcached.
+        Remcached will support rich data structure like list/map/set,
```
Comment (LINKIWI) [inline pymemcache/client/base.py]: This class need not be aware of potential subclass implementations. Same as above, let's remove the reference here.

File: pymemcache/client/base.py [hunk 5]
```diff
@@ -739,6 +744,22 @@ def _raise_errors(self, line, name):
             error = line[line.find(b' ') + 1:]
             raise MemcacheServerError(error)
 
+    def _key_helper(self, key, remapped_keys, prefixed_keys):
```
Comment (LINKIWI) [inline pymemcache/client/base.py]: While we do need to eventually override this method's logic, by itself it is not particularly useful. It may be more useful to extract the logic under `elif line.startswith(b'VALUE'):` to a separate method. Can you elaborate on the issues you ran into when you tried to do that?

From a cursory glance seems like you can pass in `remapped_keys`, `prefixed_keys`, and `line` as inputs and keep most of the logic intact.

File: pymemcache/client/base.py [hunk 6]
```diff
@@ -293,7 +293,8 @@ def close(self):
             finally:
                 self.sock = None
 
-    def set(self, key, value, expire=0, noreply=None):
+    def set(self, key, value, expire=0, noreply=None, flag=None,
+            encoding='ascii'):
```
Comment (jparise) [inline pymemcache/client/base.py]: It's a best practice to add new keyword arguments to the end of the signature in case someone has made an assumption about the positional order.

Also, I think this should be spelled `flags` to match the memcached protocol. And should this default to `0` instead of `None`?

File: pymemcache/client/base.py [hunk 7]
```diff
@@ -304,6 +305,9 @@ def set(self, key, value, expire=0, noreply=None):
                   from the cache, or zero for no expiry (the default).
           noreply: optional bool, True to not wait for the reply (defaults to
                    self.default_noreply).
+          flag: memcached operation flag
```
Comment (jparise) [inline pymemcache/client/base.py]: (Let's swap the order of `flags` and `encoding` here, too.

We should also provide a little more document borrowed from the protocol:

https://github.com/memcached/memcached/blob/78eb7701e0823643d693c1a7a6fd8a0c75db74d8/doc/protocol.txt#L175-L181

Something like:

```
flags: int, arbitrary bit field used for server-specific flags
```

File: pymemcache/client/base.py [hunk 8]
```diff
@@ -739,6 +744,32 @@ def _raise_errors(self, line, name):
             error = line[line.find(b' ') + 1:]
             raise MemcacheServerError(error)
 
+    def _value_helper(self, expect_cas, line, buf, remapped_keys,
+                      prefixed_keys):
+        """
+        This function is abstracted from _fetch_cmd to support different ways
+        of value extraction. In order to use this feature, _value_helper needs
+        to be overriden in the subclass.
+
```
Comment (jparise) [inline pymemcache/client/base.py]: nit: extra newline

File: pymemcache/client/base.py [hunk 9]
```diff
@@ -755,30 +786,16 @@ def _fetch_cmd(self, name, keys, expect_cas):
             buf = b''
             result = {}
             while True:
+
```
Comment (jparise) [inline pymemcache/client/base.py]: nit: extra newline

File: pymemcache/client/base.py [hunk 10]
```diff
@@ -755,30 +786,16 @@ def _fetch_cmd(self, name, keys, expect_cas):
...
-                    key = remapped_keys[key]
-
-                    if self.deserializer:
-                        value = self.deserializer(key, value, int(flags))
-
-                    if expect_cas:
-                        result[key] = (value, cas)
-                    else:
-                        result[key] = value
+                    key, value, buf \
```
Comment (jparise) [inline pymemcache/client/base.py]: `\`-based continuations are discouraged. Please wrap something like this:

```python
key, value, buf = self._value_helper(expect_cas, line, buf,
                                     remapped_keys, prefixed_keys)
```

File: pymemcache/client/base.py [hunk 11]
```diff
@@ -794,7 +811,8 @@ def _fetch_cmd(self, name, keys, expect_cas):
                 return {}
             raise
 
-    def _store_cmd(self, name, values, expire, noreply, cas=None):
+    def _store_cmd(self, name, values, expire, noreply, cas=None, flag=None,
```
Comment (jparise) [inline pymemcache/client/base.py]: `flag` -> `flags` and move to last argument.

File: pymemcache/client/base.py [hunk 12]
```diff
@@ -803,7 +821,7 @@ def _store_cmd(self, name, values, expire, noreply, cas=None):
             extra += b' ' + cas
         if noreply:
             extra += b' noreply'
-        expire = six.text_type(expire).encode('ascii')
+        expire = six.text_type(expire).encode(encoding)
```
Comment (jparise) [inline pymemcache/client/base.py]: This looks like a good (but separate) bug fix. Can you submit it as its own Pull Request so we can easily give it its own tests and changelog entry?

(including the related lines below)

File: pymemcache/client/base.py [hunk 13]
```diff
@@ -813,19 +831,23 @@ def _store_cmd(self, name, values, expire, noreply, cas=None):
             if self.serializer:
                 data, flags = self.serializer(key, data)
             else:
-                flags = 0
+                # flag is passed only by remcached operations
```
Comment (jparise) [inline pymemcache/client/base.py]: Please remove this comment. It's not generally applicable.

File: pymemcache/client/base.py [hunk 14]
```diff
@@ -813,19 +831,23 @@ def _store_cmd(self, name, values, expire, noreply, cas=None):
...
+                if flag:
```
Comment (jparise) [inline pymemcache/client/base.py]: If we adopt `flags`, then bit of code can just use the `flags` parameter directly.

File: pymemcache/client/base.py (hunk 12 above)
Comment (JianGuoPinterest) [reply pymemcache/client/base.py]: remcached only supports text protocol but not binary protocol, in order for the remcached extension to consume pymemcahe, the encoding has to be changed to 'utf-8', that's why I added param encoding. Do you think it's necessary to have a separate PR?

File: pymemcache/client/base.py (hunk 6 above)
Comment (jparise) [reply pymemcache/client/base.py]: Please ignore the first part of my comment. I mistakenly thought that `encoding` already existed here.

File: pymemcache/client/base.py (hunk 12 above)
Comment (jparise) [reply pymemcache/client/base.py]: Got it. I _do_ think it would be helpful to break this into three smaller (sequential) PRs:

1. The `encoding` addition.
2. The `flags` addition.
3. The "value helper".

Comment (JianGuoPinterest) [reply pymemcache/client/base.py]: totally! Will do!

File: pymemcache/client/base.py [hunk 15]
```diff
@@ -304,6 +304,7 @@ def set(self, key, value, expire=0, noreply=None):
                   from the cache, or zero for no expiry (the default).
           noreply: optional bool, True to not wait for the reply (defaults to
                    self.default_noreply).
+          encoding: determines how the data is encoded(binary or text).
```
Comment (jparise) [inline pymemcache/client/base.py]:
```suggestion
          encoding: optional str, controls data encoding (defaults to 'ascii').
```

**Comment by LINKIWI** [comment]
The smaller-scope change looks pretty good to me. As jparise said let's add a test case for a non-default encoding, after which this will probably be good to merge.

**Comment by JianGuoPinterest** [comment]
> Should we also add `encoding` to `set_many()`?
 
On remcached side, we do not consume the set_many() function, so logically we don't have to.

**Comment by jparise** [comment]
> > Should we also add `encoding` to `set_many()`?
> 
> On remcached side, we do not consume the set_many() function, so logically we don't have to.

I think we should add it to `set_many()` for consistency, especially because it just calls through to the existing support in `_store_cmd()`.

File: pymemcache/client/base.py [hunk 16]
```diff
@@ -794,7 +796,8 @@ def _fetch_cmd(self, name, keys, expect_cas):
                 return {}
             raise
 
-    def _store_cmd(self, name, values, expire, noreply, cas=None):
+    def _store_cmd(self, name, values, expire, noreply, cas=None,
+                   encoding='ascii'):
```
Comment (jparise) [inline pymemcache/client/base.py]: Because `_store_cmd()` is an internal method, I think we should move `encoding` just before `cas` and _not_ give it a default value (so no keyword argument) because we're always providing a value for it from the public methods.

Comment (JianGuoPinterest) [reply pymemcache/client/base.py]: _store_cmd() is being called by add()/ append()/ cas()/ prepend()/ replace()/ set()/ set_many(), from what you mentioned above, in order to make it consistency, should we add encoding for all these functions?

Comment (jparise) [reply pymemcache/client/base.py]: > _store_cmd() is being called by add()/ append()/ cas()/ prepend()/ replace()/ set()/ set_many(), from what you mentioned above, in order to make it consistency, should we add encoding for all these functions?

Eek, you're right. Should `encoding` be a Client-wide setting (set in `__init__()`)? Or might we vary it based on each `set` operation?

Comment (JianGuoPinterest) [reply pymemcache/client/base.py]: I'm pretty sure across remcached usage, we only use text protocol which is 'utf-8'. if we add encoding to `__init__()`, then in set and _store_cmd function, we won't need to add parameter encoding. Since the default encoding will be 'ascii', I don't think adding encoding into `__init__()` could cause any trouble for the current usage. If you think it's a good solution, i'll give it a try.

Comment (jparise) [reply pymemcache/client/base.py]: > I'm pretty sure across remcached usage, we only use text protocol which is 'utf-8'. if we add encoding to `__init__()`, then in set and _store_cmd function, we won't need to add parameter encoding. Since the default encoding will be 'ascii', I don't think adding encoding into `__init__()` could cause any trouble for the current usage. If you think it's a good solution, i'll give it a try.

Assuming we're only going to apply this to value (and not key) encoding, I think setting this in `__init__()` will work well.

Keys are expected to be ASCII-encoded so that shouldn't be configurable: https://github.com/memcached/memcached/blob/78eb7701e0823643d693c1a7a6fd8a0c75db74d8/doc/protocol.txt#L612-L615

Comment (JianGuoPinterest) [reply pymemcache/client/base.py]: yes, it's only for values. I'll work on it now.

File: pymemcache/test/test_client.py [hunk 17]
```diff
@@ -131,9 +131,21 @@ def test_set_success(self):
...
+        # unit test for encoding passed in __init__()
+        client = self.make_client([b'STORED\r\n'], encoding='utf-8')
+        result = client.set(b'key', b'value', noreply=False)
+        assert result is True
+
     def test_set_future(self):
         client = self.make_client([b'STORED\r\n'])
-        result = client.set(newbytes(b'key'), newbytes(b'value'), noreply=False)
+        result = client.set(newbytes(b'key'), newbytes(b'value'),
+                            noreply=False)
```
Comment (jparise) [inline pymemcache/test/test_client.py]: I think, if you move `noreply` back to the pervious line as it was before, we'll get a cleaner diff here (and there isn't a reason to break this line here).
